h2o-core/src/main/java/water/util/FrameUtils.java

```diff
@@ -334,25 +339,64 @@ public  double weightedMean() {
     final boolean _overwrite;
     final Job _j;
     int _nParts;
+    boolean _parallel;
     final CompressionFactory _compressor;
     final Frame.CSVStreamParams _csv_parms;
 
-    public ExportTaskDriver(Frame frame, String path, String frameName, boolean overwrite, Job j, int nParts,
+    public ExportTaskDriver(Frame frame, String path, String frameName, boolean overwrite, Job j, 
+                            int nParts, boolean perChunk,
                             CompressionFactory compressor, Frame.CSVStreamParams csvParms) {
       _frame = frame;
       _path = path;
       _frameName = frameName;
       _overwrite = overwrite;
       _j = j;
       _nParts = nParts;
+      _parallel = perChunk;
       _compressor = compressor;
       _csv_parms = csvParms;
     }
 
     @Override
     public void compute2() {
       _frame.read_lock(_j._key);
-      if (_nParts == 1) {
+      if (_parallel && _nParts == 1) {
+        _nParts = _frame.anyVec().nChunks();
+        int processed = 0;
+        String compression = H2O.getSysProperty("export.csv.cache.compression", "none");
+        CompressionFactory compressor = CompressionFactory.make(compression);
+        DecompressionFactory decompressor = DecompressionFactory.make(compression);
+        String cacheStorage = H2O.getSysProperty("export.csv.cache.storage", "memory");
+        CsvChunkCache cache = "memory".equals(cacheStorage) ? new DkvCsvChunkCache() : new FileSystemCsvChunkCache();
+        Log.info("Using compression=`" + compressor.getName() + 
+                "` and cache=`" + cache.getName() + "` for interim partial CSV export files.");
+        ChunkExportTask chunkExportTask = cache.makeExportTask(_frame, _csv_parms, compressor);
+        H2O.submitTask(new LocalMR(chunkExportTask, H2O.NUMCPUS));
+        try (FileOutputStream os = new FileOutputStream(_path)) {
+          byte[] buffer = new byte[BUFFER_SIZE];
+          boolean[] isChunkCompleted = new boolean[_nParts + 1];
+          while (processed != _nParts) {
+            int cid = chunkExportTask._completed.take();
+            isChunkCompleted[cid] = true;
+            while (isChunkCompleted[processed]) {
+              try (InputStream rawInputStream = cache.getChunkCsvStream(chunkExportTask, processed);
+                   InputStream is = decompressor.wrapInputStream(rawInputStream)) {
+                IOUtils.copyLarge(is, os, buffer);
+              } finally {
+                cache.releaseCache(chunkExportTask, processed);
+              }
+              processed++;
+              _j.update(1);
+            }
+          }
+        } catch (IOException e) {
+          throw new RuntimeException("File export failed", e);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException("File export failed", e);
```

Comment:
The interruption from the stacktrace would be obvious anyway, but stil...

```suggestion
          throw new RuntimeException("File export interrupted", e);
```

Reply:
thank you, fixed